Experimental serializer

TH3CHARLie opened the pull request:
An attempt to add serialization/deserialization to the Halide frontend. Using flatbuffers (https://flatbuffers.dev/) as the binary format.

Work in progress right now, will summarize all unsupported features/design decisions in a bit.

Shout out to @zvookin 's previous attempt (https://github.com/halide/Halide/compare/main...halide_ir_flatbuffer) which I take heavy reference from.

steven-johnson commented on apps/serdes/Deserializer.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "halide_ir_generated.h"
+#include <Halide.h>
+using namespace Halide;
```
It is *never* ok to do `using namespace Whatever;` in a .h file; it means that everything that includes this file will get this namespace opened, which can make for awful errors sometimes. Spell out all namespaces in header files.

(It's occasionally ok to do `using namespace` in `.cpp` files -- we do this frequently in tests -- but aside from tests, avoid this entirely; either do `using Halide::Internal::Whatever;` or, in some cases, `using alias = Really::Long::Namespace::Sequence;`)

TH3CHARLie replied:
Thanks for pointing this out!

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+#include <fstream>
+#include <iostream>
+
+std::string Deserializer::deserialize_string(const flatbuffers::String *str) {
```
If `str` can't ever (legally) be null, we should pass it by const-ref rather than const-ptr.

steven-johnson commented on apps/serdes/Deserializer.h:
```diff
@@ -0,0 +1,32 @@
+    // helper functions to deserialize each type of object
+    std::string deserialize_string(const flatbuffers::String *str);
+
+    Halide::Type deserialize_type(const Halide::Serdes::Type *type);
```
I assume that `Serdes` is short for `Serialize/Deserialize`, but, uh, it's not a great name. I'd vote for `Serialize`; it seems pretty obvious that deserialization is included. But I'm open for other suggestions.

TH3CHARLie replied:
I agree, `Serialize` reads good

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+
+    halide_type_code_t code = halide_type_uint;
+    switch (code_deserialized) {
+    case Halide::Serdes::TypeCode::TypeCode_Int:
```
You could make these switches easier to read by putting `using Halide::Serdes::TypeCode;` at the top of this file, then just referencing `TypeCode` directly here (and elsewhere).

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+}
+
+Halide::Type Deserializer::deserialize_type(const Halide::Serdes::Type *type) {
+    // bits
```
Here and elsewhere: these sort of comments don't add anything that isn't obvious from the code itself. Yes, we are getting `bits`, but I could have guessed that from the next line :-)

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+
+    // output_types
+    std::vector<Halide::Type> output_types;
+    auto output_types_serialized = function->output_types();
```
Possibly-ignorant question (since I don't know Flatbuffers well): do we need to make a copy here just to do a read-only iteration? ie, can we just do `for (const auto &type : *function->output_types()) {` instead?

TH3CHARLie replied:
We can! I believe this line was written when I was still learning how to use the flatbuffers API, you can see I wrote read-only iteration style code in later code when deserializing other structs.

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+    return Halide::Internal::Function(name, origin_name, output_types, required_types, required_dim, args);
+}
+
+Halide::Internal::Stmt Deserializer::deserialize_stmt(uint8_t type_code, const void *stmt) {
```
Seems like the `type_code` argument should really be `Halide::Serdes::TypeCode` instead. Why are we using `uint8` here?

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+    case Halide::Serdes::Stmt_Realize: {
+        const Halide::Serdes::Realize *realize_stmt = (const Halide::Serdes::Realize *)stmt;
+        auto name = deserialize_string(realize_stmt->name());
+        std::vector<Type> types;
```
Here and elsewhere: if we can get the `size()` of the Flatbuffers array we are reading from, we should call `.reserve()` on the `vector` we create, to reduce memory resizing churn, e.g.

```
        std::vector<Type> types;
        types.reserve(realize_stmt->types().size());  // Will this work?
```

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+        return Halide::Internal::Atomic::make(producer_name, mutex_name, body);
+    }
+    default:
+        std::cerr << "unknown type code " << type_code << "\n";
```
Don't write to `stderr` directly from within libHalide (some usages bottleneck our error output, e.g. in the Python bindings). 

If this is a fatal error, use `user_assert()`; if it's just recoverable (but needs a warning), use `user_warning`.

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+    case Halide::Serdes::Expr::Expr_IntImm: {
+        const Halide::Serdes::IntImm *int_imm_expr = (const Halide::Serdes::IntImm *)expr;
+        auto value = int_imm_expr->value();
+        // TODO: fix this hard-coding
```
What is the hard-coding in question here? The comment isn't clear enough.

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+        const Halide::Serdes::Shuffle *shuffle_expr = (const Halide::Serdes::Shuffle *)expr;
+        auto vectors_type = shuffle_expr->vectors_type();
+        auto vectors_serialized = shuffle_expr->vectors();
+        std::vector<Halide::Expr> vectors;
```
Here and elsewhere: seems like writing a `deserialize_expr_vector()` method would pay for itself.

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+    // unpack binary file
+    std::ifstream in(filename, std::ios::binary | std::ios::in);
+    if (!in) {
+        std::cerr << "failed to open file " << filename << "\n";
```
See previous comment about stderr and error reporting

steven-johnson commented on apps/serdes/Deserializer.cpp:
```diff
@@ -0,0 +1,608 @@
+    in.seekg(0, std::ios::end);
+    int size = in.tellg();
+    in.seekg(0, std::ios::beg);
+    char *data = new char[size];
```
This is never freed. Does `GetPipeline()` take ownership below? If not, this is a leak.

If GetPipeline *does* take ownership, that needs a comment.

If GetPipeline *doesn't* take ownership, it would be cleaner to do something like

```
std::vector data(size);
in.read(data.data(), size);
```

TH3CHARLie replied:
Thanks for finding this!

steven-johnson commented on apps/serdes/Printer.h:
```diff
@@ -0,0 +1,27 @@
+#include <Halide.h>
+using namespace Halide;
+
+class Printer {
```
This class would be of much greater utility if you were able to pass it a `std::ostream&` instead of having it assume `stdout`.

steven-johnson commented on apps/serdes/Serializer.h:
```diff
@@ -0,0 +1,35 @@
+#include <vector>
+
+#include "halide_ir_generated.h"
+#include <Halide.h>
```
See previous comment about <Halide.h>

steven-johnson commented on apps/serdes/halide_ir.fbs:
```diff
@@ -0,0 +1,416 @@
+file_identifier "HLD0";
+
+// File extension of any written files.
+file_extension "hir";
```
I'm confused: elsewhere, we use `.hlb` as the extension?

steven-johnson commented on apps/serdes/halide_ir.fbs:
```diff
@@ -0,0 +1,416 @@
+namespace Halide.Serdes;
+
+// This corresponds to the version.
+file_identifier "HLD0";
```
Is the identifier required to be exactly four characters?

TH3CHARLie replied:
Yes, from flatbuffers' docs:
```
Now, you can always prefix a FlatBuffer with your own file header, but FlatBuffers has a built-in way to add an identifier to a FlatBuffer that takes up minimal space, and keeps the buffer compatible with buffers that don't have such an identifier.

You can specify in a schema, similar to root_type, that you intend for this type of FlatBuffer to be used as a file format:

file_identifier "MYFI";
Identifiers must always be exactly 4 characters long. These 4 characters will end up as bytes at offsets 4-7 (inclusive) in the buffer.
```
https://flatbuffers.dev/md__schemas.html

steven-johnson commented on apps/serdes/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# ignore halide serialized buffer files
+*.hlb
```
Not at all important right now, but it's been a long time since ~any OS we care about was limited to three-character file extensions; something a bit less terse might be nice. (e.g., TFLite uses `.tflite` as an extension, which is a lot easier to read IMHO)

TH3CHARLie replied:
I propose using `.hlpipe` (Halide pipeline)

steven-johnson commented on apps/serdes/Printer.h:
```diff
@@ -0,0 +1,27 @@
+public:
+    Printer() = default;
+
+    void print_pipeline(const Halide::Pipeline &pipeline);
```
I'm a little confused about the purpose of this class: it appears to be designed to emit human-readable text (so it's not useful for any serialization/deserialization purpose)... but we already have something like that in Halide in the `IRPrinter` class. Was it written just to help with your development effort, or does it have a use that I am overlooking?

TH3CHARLie replied:
it's mainly for testing purposes. It does a pretty similar job as `IRPrinter` but supports more structs like `Bound` and `StorageDim`. Right now I use it to carry out roundtrip test before/after serialization as I'm still uncertain about the correct test strategy

steven-johnson commented on apps/serdes/Serializer.cpp:
```diff
@@ -0,0 +1,755 @@
+        auto extent = for_stmt->extent;
+        auto extent_serialized = serialize_expr(builder, extent);
+        Halide::Serdes::ForType for_type = Halide::Serdes::ForType::ForType_Serial;
+        switch (for_stmt->for_type) {
```
(Here and elsewhere:)

Using a switch statement like this to ensure that the serialized values match the compiled-into-code values is absolutely the safest thing to do to ensure we don't end up with embarrassing breakages if these get out of sync.

That said, as a bit of premature optimization... what if we did something like

```
static_assert((int)Halide::Internal::ForType::Serial == (int)Halide::Serdes::ForType::ForType_Serial);
static_assert((int)Halide::Internal::ForType::Parallel == (int)Halide::Serdes::ForType::ForType_Parallel);
... etc for the rest
for_type = (Halide::Serdes::ForType)(int)for_stmt->for_type;
```

...ewww. After having typed that out, it's definitely ugly and I'd avoid it unless I had profiling data that showed it was a meaningful performance improvement.

steven-johnson commented on apps/serdes/Serializer.cpp:
```diff
@@ -0,0 +1,755 @@
+        case Halide::Internal::ForType::GPULane: {
+            for_type = Halide::Serdes::ForType::ForType_GPULane;
+            break;
+        }
```
Here and elsewhere: how do we ensure that additions to one of the enums (in either the `.fbs` or the Halide headers files) are correctly handled? I don't know the answer to this (and I'm not expecting you to have the answer, either), but I hope there exists an answer that is more robust than "hope we don't forget someday".

TH3CHARLie replied:
I think we can alleviate this by removing the default case of the switch statement, and do the error handling outside of the switch. Then whenever we have a new addition to the enums in Halide headers (I think the workflow will be starting with adding to Halide files and then serialization protocol `fbs` file), the compiler will give an error here because of non-exhaustive enumeration

steven-johnson commented on apps/serdes/halide_ir.fbs:
```diff
@@ -0,0 +1,416 @@
+    Atomic,
+}
+
+// TODO: handle type?
```
I don't understand this comment

steven-johnson commented on src/Function.h:
```diff
@@ -68,6 +68,10 @@ class Function {
     /** Construct a Function from an existing FunctionContents pointer. Must be non-null */
     explicit Function(const FunctionPtr &);
 
+    /** Construct a function from deserializing */
```
The arguments to this need documentation:
- How is origin_name different from name?
- How are required_types different from output_types?
- What kind of strings are ok for args?
- What happens if I pass empty vectors for any of the arguments?

steven-johnson reviewed (commented):
This looks like a really nice start! My comments are mostly regarding code style and structure. I suspect @zvookin will have more substantive comments about the Flatbuffer usage, as he has far more experience than I with that.

TH3CHARLie commented:
@steven-johnson Thanks for the review comments! I'm still merging some code related to function schedules, I'll have a commit later addressing the comments.

TH3CHARLie commented:
I'm closing in on finishing implementing the core part, with all IR structs starting top-down from `Pipeline` having been implemented. I want to get feedback at/after this Friday's dev meeting. A round of review will be highly appreciated and helpful. @steven-johnson @zvookin  @abadams @derek-gerstmann 

The code is very structural and repetitive so shouldn't be too hard to review, but there are some review points:
- `FunctionPtr` is not directly serialized, it first turns into a mapping from function ptr to integer index and all function ptrs turn into an index and this index gets serialized. Later this mapping's reverse mapping gets reconstructed during deserialization and a function ptr is rebuilt. To achieve this, we first construct empty functions to get their ptrs and update them with deserialized fields (that's why we need `update_with_deserialization`)
- For union data like `Expr` and `Stmt`, it will need an extra type_code to know which union member it is, so you will see a lot of `deserialize(foo->bar_type(), foo->bar())` code. 
- any changes in `src/` are for exposing the IR to serialization, no other purposes. Not very sure about their design though so very happy to change the implementation.

What's not there and I'll implement/fix them very soon:
- [ ] testing. I'll intercept the lowering process of a pipeline, do a serialization round-trip, and see if tests can still pass. This can be part of the test suite later.
- [ ] self-reference function's weak reference is not addressed
- [ ] `Expr`'s `Type` is not properly handled, causing some hard coding right now.

derek-gerstmann commented on apps/serdes/serialize_test.cpp:
```diff
@@ -0,0 +1,25 @@
+#include <Halide.h>
+using namespace Halide;
+
+// a simple round trip test
```
What if we had an IR comparator to validate the deserialized pipeline matches the original in-memory pipeline?

TH3CHARLie replied:
See `IRComparator.cpp`

TH3CHARLie commented on apps/serdes/Deserializer.h:
```diff
@@ -0,0 +1,99 @@
+    Halide::Pipeline deserialize(const std::string &filename);
+
+private:
+    std::map<int32_t, Halide::Internal::FunctionPtr> reverse_function_mappings;
```
no, we don't need its elements sorted

TH3CHARLie replied:
updated

TH3CHARLie commented on apps/serdes/Serializer.cpp:
```diff
@@ -0,0 +1,1106 @@
+}
+
+std::pair<Halide::Serialize::Expr, flatbuffers::Offset<void>> Serializer::serialize_expr(flatbuffers::FlatBufferBuilder &builder, const Halide::Expr &expr) {
+    if (!expr.defined()) {
```
I'm not sure about that, actually with the example code in the current `serialize_test` it's going to have undefined `Expr` or `Stmt` (that's how I find out about this). There could be empty `Expr` in places like `Func::extern_proxy_expr` that just got default constructed so it's undefined.

TH3CHARLie commented on apps/serdes/halide_ir.fbs:
```diff
@@ -0,0 +1,678 @@
+
+// File extension of any written files. "hlpipe" stands for Halide Pipeline.
+file_extension "hlpipe";
+
```
good point, this is a good approach to prevent IR update breaks serialization.

TH3CHARLie replied:
The best we can do as far as I tried is to use the `file_identifier` field in serialization format, for example for Halide 17.0 we use "HL17"

steven-johnson replied:
Right, but I think Derek is suggesting we add another struct with explicitly-populated metadata that may be useful / necessary for future consumers.

TH3CHARLie replied:
oh I see, we can add that in the Pipeline like 
```
table Pipeline {
...
version: string;
}
```
with the string being something like "Halide v17.0.0-CHKSUM"?

steven-johnson replied:
Yep. Though I suspect it might be better as a struct since we may well want to extend it, eg

table VersionMetadata {
   halide_version: string; // major.minor.patch
   halide_commit; string;  // git commit of Halide release
   halide_target: string;  // Target used for generation, or empty string if Target() used
}

TH3CHARLie commented:
@steven-johnson Right now the status of this PR is that we are testing serialization using the `correctness` suite, with the most recent commit reaching 90% passed tests (the remaining ones are related to the jit handlers because we are doing jit compilation interception). Hopefully I'll resolve the jit handler issue and make serialization so-called thoroughly tested. Then this PR should be ready-to-review, with expected-to-come build issues to discuss of course.

About serializing the schedule part, that is a great idea and sounds not that hard to achieve, I'll look into that once this lands.

steven-johnson commented on src/Definition.cpp:
```diff
@@ -108,6 +108,18 @@ Definition::Definition(const std::vector<Expr> &args, const std::vector<Expr> &v
     }
 }
 
+Definition::Definition(bool is_init, const Expr &predicate, const std::vector<Expr> &args, const std::vector<Expr> &values,
```
Code style nit: prefer to avoid bool args to functions like this; it's hard to read at the call site, because the semantic of what it means isn't clear. Prefer to use something like `enum class DefinitionInit { IsInit, IsNotInit};` so that the meaning at the call site is immediately obvious.

TH3CHARLie replied:
In this case at the call sites, it's just treating the boolean as another field of the deserialized `Definition` and it doesn't have any additional semantic meaning other than that, so I'm not sure if we want to do the enum change here.

steven-johnson replied:
Ah, gotcha. OK.

steven-johnson commented on src/Deserialization.h:
```diff
@@ -0,0 +1,10 @@
+#include "Pipeline.h"
+
+namespace Halide {
+Pipeline deserialize_pipeline(const std::string &filename, const std::unordered_map<std::string, Internal::Parameter> &params);
```
Does this need any comments about argument restrictions, error conditions, etc? (Not sure, this is a genuine question)

steven-johnson commented on src/Function.cpp:
```diff
@@ -331,6 +331,35 @@ Function::Function(const std::vector<Type> &required_types, int required_dims, c
     contents->required_dims = required_dims;
 }
 
+void Function::update_with_deserialization(const std::string &name, const std::string &origin_name, const std::vector<Halide::Type> &output_types,
```
...wow, this has a lot of arguments.
(1) Code style nit: for this many args, I'd argue that we should really put one per line (both here and in the header), as it's hard to read and check, IMHO)
(2) I wonder if this would be better as a struct argument: done well, that would allow some of probably-lesser-used args (e.g. debug_file) to have default values that don't have to be filled in. OTOH, I guess the 'struct' would really be basically just the entire FunctionContents, which is intentionally hidden.

Maybe I'm overthinking point 2 above; I suspect this function will ~never be called by user code directly (only indirectly via the deserialization code).

steven-johnson commented on src/Parameter.cpp:
```diff
@@ -84,6 +79,26 @@ Parameter::Parameter(const Type &t, bool is_buffer, int d, const std::string &na
+    contents->buffer_constraints = buffer_constraints;
+    contents->memory_type = memory_type;
+}
+Parameter::Parameter(const Type &t, bool is_buffer, int dimensions, const std::string &name,
```
style nit: blank line between functions

steven-johnson commented on src/Parameter.h:
```diff
@@ -55,6 +60,15 @@ class Parameter {
      * explicitly specified (as opposed to autogenerated). */
     Parameter(const Type &t, bool is_buffer, int dimensions, const std::string &name);
 
+    /** Construct a new parameter via deserialing. */
```
I don't think `deserialing` is a word. Deserializing? Deserialization?

steven-johnson commented on src/Pipeline.cpp:
```diff
@@ -555,6 +566,21 @@ Target Pipeline::get_compiled_jit_target() const {
+        debug(0) << "intercept compile_jit, serialization completed\n";
+        deserialized_pipe = deserialize_pipeline(filename, call_params);
+        debug(0) << "intercept compile_jit, deserialization completed\n";
+        // deserialized_pipe.jit_handlers() = this->jit_handlers();
```
As always, commented out code should be removed or have a comment, even for a draft PR :-)

(eg even just `// TODO: commented out pending foo`)

steven-johnson commented on src/Pipeline.cpp:
```diff
@@ -555,6 +566,21 @@ Target Pipeline::get_compiled_jit_target() const {
+    Pipeline deserialized_pipe;
+    if (!debug_serialization.empty()) {
+        std::string filename = generate_function_name() + ".hlpipe";
+        debug(0) << "intercept compile_jit, file: " << filename << "\n";
```
I assume that the `debug(0)` will be either removed or converted to `debug(1)` or 2 or 3, etc before landing?

steven-johnson commented on src/Pipeline.cpp:
```diff
@@ -565,25 +591,54 @@ void Pipeline::compile_jit(const Target &target_arg) {
-
+    if (!debug_serialization.empty()) {
+        deserialized_pipe.infer_arguments();
+        // infer_arguments();
```
As always, commented out code should be removed or have a comment, even for a draft PR :-)

steven-johnson commented on src/Pipeline.cpp:
```diff
@@ -565,25 +591,54 @@ void Pipeline::compile_jit(const Target &target_arg) {
+            args.push_back(arg.arg);
+        }
+        contents->inferred_args = deserialized_pipe.get_inferred_args();
+        // if (this->jit_handlers().custom_error != nullptr) {
```
As always, commented out code should be removed or have a comment, even for a draft PR :-)

steven-johnson commented on src/Pipeline.cpp:
```diff
@@ -565,25 +591,54 @@ void Pipeline::compile_jit(const Target &target_arg) {
+            outputs.push_back(f.function());
+        }
+        contents->jit_cache = deserialized_pipe.compile_jit_cache(module, std::move(args), outputs, deserialized_pipe.get_jit_externs(), target);
+        // deserialized_pipe.compile_jit_cache()  = contents->jit_cache;
```
As always, commented out code should be removed or have a comment, even for a draft PR :-)

steven-johnson commented on src/Pipeline.cpp:
```diff
@@ -565,25 +591,54 @@ void Pipeline::compile_jit(const Target &target_arg) {
     std::map<std::string, JITExtern> lowered_externs = contents->jit_externs;
-    contents->jit_cache = compile_jit_cache(module, std::move(args), contents->outputs, contents->jit_externs, target);
+    if (!debug_serialization.empty()) {
+        std::vector<Function> outputs;
```
...my kingdom for a real list-comprehension syntax in C++ :-/

TH3CHARLie replied:
I'm not sure about this comment...

steven-johnson commented on src/Deserialization.h:
```diff
@@ -0,0 +1,10 @@
+
+#include "Pipeline.h"
+
+namespace Halide {
```
Style nit: blank lines before and after here

steven-johnson commented on src/halide_ir.fbs:
```diff
@@ -0,0 +1,692 @@
+    func_names_in_order: [string];
+    parameters: [Parameter];
+    buffers: [Buffer];
+    // trace_pipeline: bool;
```
As always, commented out code should be removed or have a comment, even for a draft PR :-)

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    Pipeline deserialize(const std::string &filename);
+
+private:
+    std::unordered_map<int32_t, FunctionPtr> reverse_function_mappings;
```
All these data members deserve at least a one-line comment.

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+public:
+    Deserializer() = default;
+
+    Deserializer(const std::unordered_map<std::string, Parameter> &params)
```
A single-argument constructor should usually be tagged as `explicit`. (If it is deliberately not explicit because you want implicit conversion, then add a `/*not-explicit*/` as a comment.)

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    case Serialize::DeviceAPI::DeviceAPI_WebGPU:
+        return DeviceAPI::WebGPU;
+    default:
+        user_assert(false) << "unknown device api " << device_api << "\n";
```
Style nit: here and elsewhere, just use `user_error` instead of `user_assert(false)`.

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+}
+
+ExternFuncArgument::ArgType Deserializer::deserialize_extern_func_argument_type(const Serialize::ExternFuncArgumentType extern_func_argument_type) {
+    switch (extern_func_argument_type) {
```
Doing all of these via switch-with-default is probably the right way to do this, but the premature optimizer in my brain wants to set up some data-driven templated table nonsense. I'm glad you resisted that urge :-)

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    }
+}
+
+Type Deserializer::deserialize_type(const Serialize::Type *type) {
```
Possibly-dumb question: if the argument can't legally be null, why not pass a reference instead of a pointer?

TH3CHARLie replied:
some fields can possibly be null, for example, see `CreateParameter` usages in `Serialization.cpp`. For consistency across all deserialization function signatures, using pointers here seems more reasonable.

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    case TypeCode::TypeCode_BFloat:
+        code = halide_type_bfloat;
+        break;
+    }
```
Shouldn't there be a default case here?

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+Type Deserializer::deserialize_type(const Serialize::Type *type) {
+    user_assert(type != nullptr) << "deserializing a null Type\n";
+    using Serialize::TypeCode;
+    int bits = type->bits();
```
Style nit, here and a lot of other places: it's (slightly) preferable code style to define locals like this as `const` when you know they won't/can't be mutated later on; it makes the code a bit easier to read when you can immediately note that this local won't change later based on something you didn't expect.

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    }
+
+    auto func_schedule = deserialize_func_schedule(function->func_schedule());
+
```
why the blank line?

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+        auto min = deserialize_expr(for_stmt->min_type(), for_stmt->min());
+        auto extent = deserialize_expr(for_stmt->extent_type(), for_stmt->extent());
+        ForType for_type = deserialize_for_type(for_stmt->for_type());
+
```
why the blank line?

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+}
+
+Expr Deserializer::deserialize_expr(Serialize::Expr type_code, const void *expr) {
+    assert(expr != nullptr);
```
...wait, is `Serialize::Expr` a pointer type?

TH3CHARLie replied:
no, it's an enum type underneath the hood

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    assert(expr != nullptr);
+    switch (type_code) {
+    case Serialize::Expr::Expr_IntImm: {
+        const Serialize::IntImm *int_imm_expr = (const Serialize::IntImm *)expr;
```
Could just use `auto *int_imm_expr` here instead -- the type is obvious from the RHS of the statement.

TH3CHARLie replied:
TODO for myself: do the same for `Stmt` as well

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    }
+}
+
+std::vector<Expr> Deserializer::deserialize_expr_vector(const flatbuffers::Vector<uint8_t> *exprs_types, const flatbuffers::Vector<flatbuffers::Offset<void>> *exprs_serialized) {
```
style nit: long line is long, maybe break second argument onto next line

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+}
+
+Range Deserializer::deserialize_range(const Serialize::Range *range) {
+    assert(range != nullptr);
```
Some of the null checks in this file use `user_assert` (which always checks) and some use `assert` (which depends on compiler mode), is this intentional?

TH3CHARLie replied:
no, we should stick to `user_assert`

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    assert(func_schedule != nullptr);
+    auto store_level = deserialize_loop_level(func_schedule->store_level());
+    auto compute_level = deserialize_loop_level(func_schedule->compute_level());
+    std::vector<StorageDim> storage_dims;
```
The pattern of

     declare vector
     reserve vector
     populate vector by calling some fn on every instance in another vector

occurs often enough here that I wonder if it would be useful to make a templated helper func to make it terser (and to ensure that reserve() is always called). What do you think? eg

```
template<typename Src, typename Dst>
std::vector<Dst> deserialize_vector(flatbuffer_collection<Src>, std::function<Dst(Src)>);
```

TH3CHARLie replied:
I like this!

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    if (!this->reverse_function_mappings.empty()) {
+        this->reverse_function_mappings.clear();
+    }
+    int cnt = 0;
```
Just spell out `count`, it's cheap :-)

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    }
+}
+
+Pipeline Deserializer::deserialize(const std::string &filename) {
```
This would be a lot more flexible if you split it into two functions, with one taking `std::istream&` as an argument (note, *not* an `ifstream`), so that callers could use serialization to/from memory, or some other thing that doesn't map well to a POSIX file.

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    std::vector<char> data(size);
+    in.read(data.data(), size);
+    in.close();
+
```
We should probably check `in.good()` to check for errors, otherwise builds with exceptions disabled might ignore them, which would be bad

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+    in.read(data.data(), size);
+    in.close();
+
+    const auto *pipeline_obj = Serialize::GetPipeline(data.data());
```
Can this result be null?

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1358 @@
+        if (buffers_in_pipeline.find(buffer.name()) == buffers_in_pipeline.end()) {
+            buffers_in_pipeline[buffer.name()] = buffer;
+        } else {
+            user_assert(false) << "duplicate buffer " << buffer.name() << " in pipeline\n";
```
here and elsewhere: why not just do

```
            user_assert(buffers_in_pipeline.count(buffer.name()) > 0) << "duplicate buffer " << buffer.name() << " in pipeline\n";
            buffers_in_pipeline[buffer.name()] = buffer;
```

(The `else` can be elided since user_assert will never return if the condition is false, and `count()` is terser than find/end since we just care about presence. Sadly, `contains()` is C++20 and we haven't yet migrated.)

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+
+    void serialize(const Pipeline &pipeline, const std::string &filename);
+
+    const std::unordered_map<std::string, Parameter> &get_external_parameters() const {
```
(Pretty much all the comments I made for Deserialization.cpp apply here too, so I won't repeat them)

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+    }
+}
+
+flatbuffers::Offset<flatbuffers::String> Serializer::serialize_string(flatbuffers::FlatBufferBuilder &builder, const std::string &str) {
```
Style nit: is it worth putting `using flatbuffers::Offset;` at the top of this file to make all this terser? (Ditto for various other flatbuffers types)

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+        specializations_serialized.push_back(serialize_specialization(builder, specialization));
+    }
+    auto source_location_serialized = serialize_string(builder, definition.source_location());
+    return Halide::Serialize::CreateDefinition(builder, is_init, predicate_serialized.first, predicate_serialized.second, builder.CreateVector(values_types),
```
Nit here and elsewhere: we don't have an explicit line-width limit in Halide style guidelines but these are getting wide enough that I have to make the GitHub diff window fullscreen on my 4k monitor to avoid line wrapping so, uh, maybe some of these could be made a bit less wide :-)

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+void Serializer::serialize(const Pipeline &pipeline, const std::string &filename) {
+    flatbuffers::FlatBufferBuilder builder(1024);
+
+    // extract the DAG, unwarp function from Funcs
```
...I assume you mean `unwrap`?

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+        funcs_serialized.push_back(this->serialize_function(builder, entry.second));
+    }
+
+    auto outpus = pipeline.outputs();
```
...`outputs`? (Or did you mean `octopus`?)

TH3CHARLie replied:
I must be hungry when i wrote this line

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+    auto requirements_vector = builder.CreateVector(requirements_serialized);
+    auto requirements_types_vector = builder.CreateVector(requirements_types);
+
+    // For Parameters and buffers, to prevent we serialize the same object multiple times, we use a map to store the unique
```
to prevent... what? Something is missing :-)

steven-johnson commented on src/Serialization.cpp:
```diff
@@ -0,0 +1,1310 @@
+        buffers_serialized.push_back(serialize_buffer(builder, buffer.second));
+    }
+
+    auto pipeline_obj = Halide::Serialize::CreatePipeline(builder, builder.CreateVector(funcs_serialized), builder.CreateVector(output_names_serialized),
```
I'm gonna say this line is too long

steven-johnson reviewed (commented):
This looks great -- almost all my comments are style nits or suggestions for code tweaks that I think would improve readability or code terseness.

alexreinking commented on src/CMakeLists.txt:
```diff
@@ -396,6 +403,17 @@ foreach (f IN LISTS HTML_TEMPLATE_FILES)
+##
+add_library(Halide_serialization_headers OBJECT)
+add_custom_command(OUTPUT "$<SHELL_PATH:${CMAKE_CURRENT_SOURCE_DIR}/halide_ir_generated.h>"
+                   COMMAND flatc --cpp -o "$<SHELL_PATH:${CMAKE_CURRENT_SOURCE_DIR}>" "$<SHELL_PATH:${CMAKE_CURRENT_SOURCE_DIR}/halide_ir.fbs>"
```
If `flatc` is a target, then this is fine, but I'll need to look at the flatbuffers documentation to be sure. It would be a lot more typical to have `flatbuffers::flatc` or the like.

alexreinking commented on src/CMakeLists.txt:
```diff
@@ -363,6 +367,9 @@ set(HTML_TEMPLATE_FILES
     StmtToViz_stylesheet
     )
 
+set (SERIALIZATION_DEFINITION
+    halide_ir.fbs
+    )
```
```suggestion
```

Unused

alexreinking commented on src/CMakeLists.txt:
```diff
@@ -420,6 +438,16 @@ add_library(Halide
             # Including these as sources works around the need to "install" Halide_initmod
             $<TARGET_OBJECTS:Halide_initmod>
             $<TARGET_OBJECTS:Halide_c_templates>)
+if (BUILD_SERIALIZATION)
+    find_package(flatbuffers REQUIRED)
+    if (flatbuffers_FOUND)
+        target_link_libraries(Halide PRIVATE flatbuffers)
+        message(STATUS "Build with flatbuffers")
+    else ()
+        message(FATAL_ERROR "flatbuffers not found")
+    endif ()
+endif ()
```
The code above defining the custom command should be deleted. Here are the changes I made:

1. Generate in `${PROJECT_BINARY_DIR}/include`. _n.b._ test the install target; if the generated header is installed, then `packaging/CMakeLists.txt` will need to be updated to exclude it.
2. Don't check `flatbuffers_FOUND` since `REQUIRED` was passed.
3. Just attach the header to the `Halide` target. There was no `.c`/`.cpp` file to include in the "object" library anyway.
4. Only bother with `$<SHELL_PATH:...>` if it's actually necessary in practice.

```suggestion
if (BUILD_SERIALIZATION)
    find_package(flatbuffers REQUIRED)
    target_link_libraries(Halide PRIVATE flatbuffers::flatbuffers)
    
    set(fb_def "${CMAKE_CURRENT_SOURCE_DIR}/halide_ir.fbs")
    set(fb_header "${PROJECT_BINARY_DIR}/include/halide_ir_generated.h")
    add_custom_command(
        OUTPUT "${fb_header}"
        COMMAND flatbuffers::flatc --cpp -o "${PROJECT_BINARY_DIR}/include" "${fb_def}"
        DEPENDS "${fb_def}"
        VERBATIM
    )
    target_sources(Halide PRIVATE "${fb_header}")
endif ()
```

Generating it in `${PROJECT_BINARY_DIR}/include` directory will likely require an additional rule to exclude it from the installation archive.

alexreinking commented on CMakeLists.txt:
```diff
@@ -21,6 +21,9 @@ include(HalideGeneratorHelpers)
 # Build Halide as a shared lib by default, but still honor command-line settings.
 option(BUILD_SHARED_LIBS "Build shared libraries" ON)
 
+# Build serialization, disabled by default 
+option(BUILD_SERIALIZATION "Build Halide serialization" ON)
```
Misleading comment... it is _enabled_ by default.

alexreinking reviewed (changes requested):
Please apply my proposed CMake changes and let me know if something doesn't work.

alexreinking commented:
CI needs to be updated to build and install flatbuffers for clang-tidy.

steven-johnson commented:
> CI needs to be updated to build and install flatbuffers for clang-tidy.

I took the liberty of pushing a hopeful fix directly to this PR

alexreinking commented:
Here is a branch that includes a commit that fixes the CMake build.

https://github.com/CodeLinaro/Halide/commit/9c2bb30e2313b2c757372db42b09f8dfc4ffa8a7

I can't open a PR-to-PR because... reasons.

alexreinking commented:
> EDIT: in theory it's being installed, but clang-tidy still fails with

Oh god, does the Ubuntu package predate flatbuffers' CMake packaging?

TH3CHARLie commented:
I install libflatbuffers-dev via ubuntu's default packing system for CMake I'm using, it should work like in the previous commit in theory......

steven-johnson commented:
According to https://packages.ubuntu.com/source/focal/flatbuffers:

[flatbuffers-compiler](https://packages.ubuntu.com/focal/flatbuffers-compiler)
efficient cross platform serialization library - compiler
[flatbuffers-compiler-dev](https://packages.ubuntu.com/focal/flatbuffers-compiler-dev)
efficient cross platform serialization library - cmake files
[libflatbuffers-dev](https://packages.ubuntu.com/focal/libflatbuffers-dev)
efficient cross platform serialization library - development files
[libflatbuffers1](https://packages.ubuntu.com/focal/libflatbuffers1)
efficient cross platform serialization library - shared library

TH3CHARLie commented:
> Here is a branch that includes a commit that fixes the CMake build.
> 
> [CodeLinaro@9c2bb30](https://github.com/CodeLinaro/Halide/commit/9c2bb30e2313b2c757372db42b09f8dfc4ffa8a7)
> 
> I can't open a PR-to-PR because... reasons.

Thanks! This works perfectly!

TH3CHARLie commented:
clang-tidy is still running but at least it can find flatbuffers right now, turns out this is some case-sensitive stuff as the cmake file is called `FlatbuffersConfig.cmake`

TH3CHARLie commented:
looks like some of the code in the repo are not formated, whenever I run the script I'll get a lot of files reformatted that are not related to this PR.

steven-johnson commented:
> looks like some of the code in the repo is not formatted, whenever I run the script I'll get a lot of files reformatted that are not related to this PR.

Are you synced to head? We changed to using clang-tidy-16 and clang-format-16 (vs 14) a little while ago. Make sure you are synced with `main` and retry.

TH3CHARLie commented:
> > looks like some of the code in the repo is not formatted, whenever I run the script I'll get a lot of files reformatted that are not related to this PR.
> 
> Are you synced to head? We changed to using clang-tidy-16 and clang-format-16 (vs 14) a little while ago. Make sure you are synced with `main` and retry.

I was kind out-of-sync before, but the commit you added today resynced with main and I'm now using clang-format-16 for all the formatting.

TH3CHARLie commented:
`#include "halide_ir_generated.h" not found` is why clang-tidy is not happy about

TH3CHARLie commented:
@steven-johnson @derek-gerstmann @alexreinking  

I believe I've addressed every review comment above (either reflected in code updates or comments), and now ready for a new round of review. I think one remaining TODO to make this PR fully mergeable (other than incoming comments from next round review) is adding CMake file/rule to have a separate serialization test. I've tried some but still haven't figure out the right way.

For the version metadata, serializing only schedules and etc, I'd like to leave them as followup PRs.

steven-johnson commented:
> think one remaining TODO to make this PR fully mergeable (other than incoming comments from next round review) is adding CMake file/rule to have a separate serialization test

I can help with the CMake rules, but does such a test exist?

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1325 @@
+    // Helper function to deserialize a homogenous vector from a flatbuffer vector,
+    // does not apply to union types like Stmt and Expr or enum types like MemoryType
+    template<typename src, typename dst>
+    std::vector<dst> deserialize_vector(const flatbuffers::Vector<::flatbuffers::Offset<src>> *flatbuffer_vec, std::function<dst(Deserializer &, const src *)> deserialize_func) {
```
suggest adding a line break after `flatbuffer_vec`

TH3CHARLie commented:
> > think one remaining TODO to make this PR fully mergeable (other than incoming comments from next round review) is adding CMake file/rule to have a separate serialization test
> 
> I can help with the CMake rules, but does such a test exist?

The tests exist in a non-traditional way, basically we utilize the existing correctness tests and run them with a new env var `HL_TEST_SERIALIZATION=1`. You can see my attempt to do something like this in the newest commit.

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1325 @@
+};
+
+std::string Deserializer::deserialize_string(const flatbuffers::String *str) {
+    user_assert(str != nullptr) << "deserializing a null string\n";
```
Is this really a user-facing error or an internal error?

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1325 @@
+    user_assert(function != nullptr) << "deserializing a null Function\n";
+    const std::string name = deserialize_string(function->name());
+    const std::string origin_name = deserialize_string(function->origin_name());
+    const std::vector<Type> output_types = deserialize_vector<Serialize::Type, Type>(function->output_types(), &Deserializer::deserialize_type);
```
several of these lines are > 140 chars, I think a line break is in order

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1325 @@
+    const auto name_mangling = deserialize_name_mangling(function->extern_mangling());
+    const auto extern_function_device_api = deserialize_device_api(function->extern_function_device_api());
+    const auto extern_proxy_expr = deserialize_expr(function->extern_proxy_expr_type(), function->extern_proxy_expr());
+    const bool trace_loads = function->trace_loads(), trace_stores = function->trace_stores(), trace_realizations = function->trace_realizations();
```
In Halide, prefer to define/init each var on a separate line, it's easier to read:

const bool trace_loads = ...
const bool trace_stores = ...

etc

steven-johnson commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1325 @@
+    return deserializer.deserialize(in);
+}
+
+}  // namespace Halide
```
Please add trailing end-of-line char

steven-johnson commented on test/CMakeLists.txt:
```diff
@@ -105,3 +105,15 @@ if (WITH_TEST_FUZZ)
+    add_subdirectory(serialization)
+else ()
+    message(STATUS "Building serialization tests disabled")
+endif ()
```
Please end all source files with a newline

steven-johnson reviewed (approved):
Aside from a couple of nits, this looks good -- my major nit is that there are still a lot of really long lines -- we don't have an explicit line-length limit, but in practice, lines that are > 120 chars wide or so are pretty rare in existing Halide code. It's a judgement call when to use really long lines, but I found that I had to make the diff window nearly the width of my 4k monitor to get the lines to stop wrapping, so, I'm gonna say that some judicious line breaks would be appropriate.

The main thing I see lacking here is (1) testing, of course, and (2) documentation. Do we have a plan for them (either here or in a subsequent PR?)

steven-johnson commented:
> > > think one remaining TODO to make this PR fully mergeable (other than incoming comments from next round review) is adding CMake file/rule to have a separate serialization test
> > 
> > 
> > I can help with the CMake rules, but does such a test exist?
> 
> The tests exist in a non-traditional way, basically we utilize the existing correctness tests and run them with a new env var `HL_TEST_SERIALIZATION=1`. You can see my attempt to do something like this in the newest commit.

Oh... yeah, that's not optimal :-)

steven-johnson reviewed (changes requested):
Undoing approval because the HL_TEST_SERIALIZATION hack in compile_jit() etc must be removed before this can land.

TH3CHARLie commented:
ping @abadams @derek-gerstmann for test discussion

TH3CHARLie commented on src/Deserialization.cpp:
```diff
@@ -0,0 +1,1325 @@
+    // does not apply to union types like Stmt and Expr or enum types like MemoryType
+    template<typename src, typename dst>
+    std::vector<dst> deserialize_vector(const flatbuffers::Vector<::flatbuffers::Offset<src>> *flatbuffer_vec, std::function<dst(Deserializer &, const src *)> deserialize_func) {
+        user_assert(flatbuffer_vec != nullptr) << "deserializing a null vector\n";
```
I think the lines are kind of blurry: We can call it a user-error if the user provided broken/not-compatible file input, otherwise we can call it an internal error

TH3CHARLie replied:
But in both ways we should stay consistent: pick one from `user_error`/`internal_error` and use across serialization implementations.

TH3CHARLie commented:
I've removed the hack for now to unblock review process. As a source of confidence, we used the hack to test serialization on both x86-linux and m1 Mac

TH3CHARLie commented:
ping for review, I guess the current plan is to have no tests associated with this PR. Serialization code is guarded with compilation flags so shouldn't be causing damage around. We will figure out a better test plan during dev meeting probably

steven-johnson commented:
We need to ensure that the clang-tidy checks pass before we can land this.

TH3CHARLie commented:
> We need to ensure that the clang-tidy checks pass before we can land this.

absolutely, I believe all contents of serialization headers should be enclosed in #define pairs

mcourteaux commented:
@TH3CHARLie Just curious: what's the goal of this serialization? What is the use case? Is it to save JIT-compiled pipelines? As in: have the entire Halide library be on a client machine, and JIT compile the Halide pipelines and store the result for future reuse? This, as opposed to shipping AOT compiled pipelines with generators?

TH3CHARLie commented:
> What's the goal of this serialization?  What is the use case? 

In cases like autoscheduling, we are relying on header-file to distribute the schedule and manually apply the schedule to algorithms, having a standard serialized format is a more efficient approach to do this.

> Is it to save JIT-compiled pipelines? 
No right now it only serialize front end IRs of a pipeline

TH3CHARLie commented:
@mcourteaux 
> > What's the goal of this serialization?  What is the use case?
> 
> In cases like autoscheduling, we are relying on header-file to distribute the schedule and manually apply the schedule to algorithms, having a standard serialized format is a more efficient approach to do this.
> 
> > Is it to save JIT-compiled pipelines?
> 
> No right now it only serialize front end IRs of a pipeline

TH3CHARLie commented:
> We need to ensure that the clang-tidy checks pass before we can land this.

I am not so similar with the clang-tidy compilation workflow, but looks like with current rules the `flatc/include` is never built in the clang-tidy build therefore always triggering a header file not found error

TH3CHARLie commented:
I see `DefaultCostModel.cpp` is skipped because it relies on generated `cost_model.h` so I do the same for serialization cpp files.

steven-johnson commented:
Pulling this PR to my local Mac (which does not have flatbuffers installed) and attempting to build fails with

```
CMake Error at src/CMakeLists.txt:431 (find_package):
  By not providing "FindFlatbuffers.cmake" in CMAKE_MODULE_PATH this project
  has asked CMake to find a package configuration file provided by
  "Flatbuffers", but CMake did not find one.

  Could not find a package configuration file provided by "Flatbuffers" with
  any of the following names:

    FlatbuffersConfig.cmake
    flatbuffers-config.cmake

  Add the installation prefix of "Flatbuffers" to CMAKE_PREFIX_PATH or set
  "Flatbuffers_DIR" to a directory containing one of the above files.  If
  "Flatbuffers" provides a separate development package or SDK, be sure it
  has been installed.
```

We should really smarten it to avoid failure in that case; let me see if I can do so.

steven-johnson commented:
The way I decided to resolve the "Flatbuffers missing" case was to provide a more useful error message if BUILD_SERIALIZATION=ON and Flatbuffers can't be found:

```
CMake Error at src/CMakeLists.txt:433 (message):
  BUILD_SERIALIZATION is ON, but could not find Flatbuffers installed
  locally.  Either install Flatbuffers or build with BUILD_SERIALIZATION=OFF.
```

I'm also going to add a bit of information about this to the README.

steven-johnson commented:
FYI: setting BUILD_SERIALIZATION=OFF means you can't build, because `halide_ir_generated.h` isn't generated and it is unconditionally included. I'll fix that too.

TH3CHARLie commented:
> FYI: setting BUILD_SERIALIZATION=OFF means you can't build, because `halide_ir_generated.h` isn't generated and it is unconditionally included. I'll fix that too.

Guard the entire serialization cpp files with some preprocessors macros? Thanks for the help on this!

steven-johnson commented:
> > FYI: setting BUILD_SERIALIZATION=OFF means you can't build, because `halide_ir_generated.h` isn't generated and it is unconditionally included. I'll fix that too.
> 
> Guard the entire serialization cpp files with some preprocessors macros? Thanks for the help on this!

Yeah, see [15ba956](https://github.com/halide/Halide/pull/7594/commits/15ba95686b95fb2943736c1ecb58a815a7352fda) and [639ee8f](https://github.com/halide/Halide/pull/7594/commits/639ee8f93c9c7081694f84d4673fe31b7b2acaa2) for the changes I made for this, including a very minimal README addition.

I'm going to take a quick stab at adding support for using Flatbuffers via FetchContent (as we do with WABT, etc) so that we don't have to install on all buildbots (and also don't have to deal with oddball platforms like arm-32-linux that probably don't have prebuilt tools); either way we should be able to land by EOD.

TH3CHARLie commented on .github/workflows/presubmit.yml:
```diff
@@ -35,7 +35,7 @@ jobs:
           sudo apt-add-repository "deb https://apt.llvm.org/$(lsb_release -sc)/ llvm-toolchain-$(lsb_release -sc)-16 main"
           sudo apt-get update
-          sudo apt-get install llvm-16 clang-16 liblld-16-dev libclang-16-dev clang-tidy-16 ninja-build flatbuffers-compiler flatbuffers-compiler-dev libflatbuffers-dev libflatbuffers1
+          sudo apt-get install llvm-16 clang-16 liblld-16-dev libclang-16-dev clang-tidy-16 ninja-build flatbuffers-compiler flatbuffers-compiler-dev libflatbuffers-dev
```
why remove libflatbuffers1 here?

TH3CHARLie commented on src/CMakeLists.txt:
```diff
@@ -443,14 +443,14 @@ if (WITH_SERIALIZATION)
     add_custom_command(
         OUTPUT "${fb_header}"
-        COMMAND flatbuffers::flatc --cpp -o "${fb_dir}" "${fb_def}"
+        COMMAND FlatBuffers::flatc --cpp -o "${fb_dir}" "${fb_def}"
```
If I remember correctly, flatbuffer package is pretty strange here that is use `FlatBuffers` as the package name but inside the package the namespace is all lower-case `flatbuffers`

steven-johnson replied:
To make matters worse, the apt packages on my Google linux box have `FlatBuffers.cmake` but the ones from GitHub Actions are `Flatbuffers.cmake`. Oy. Gonna switch back to the way it was before since that seems to be more likely.

TH3CHARLie replied:
yikes that sounds bad

TH3CHARLie reviewed (commented):
This may help resolve the recent clang-tidy complaint

TH3CHARLie commented:
it worked!

steven-johnson commented:
Apparently the Flatbuffers versioning scheme changed radically a year or so ago... in Aug 2022, they released v2.0.8 (and this is the version that you get via `apt` on Ubuntu). The next release, in Sep 2022, was... 22.9.24 (based on the date, I guess); this is the version you get via Homebrew. I have no idea what minimum version we need. For now I guess we won't bother with a version check.

steven-johnson commented:
OK, crossing my fingers that this is the final revision -- turns out that using FetchContent for flatbuffers/flatc is arguably easier than finding an installed one, so that's what we're gonna do by default. If this is green, I'm landing it :-)

steven-johnson commented:
Update on the clang-tidy work: as of top-of-tree here, running clang-tidy on this branch works correctly on my local Linux box (which is a 48-core machine with lots of RAM); on GHA here though, it seems to stall -- maybe the new Serialization and/or generated flatbuffer headers are overwhelming clang-tidy on the more limited GHA machines? Let me see if I can turn down parallelization.

steven-johnson reviewed (approved):
Finally LGTM pending green.

steven-johnson commented:
Just waiting for @alexreinking to approve since there are pending "requested changes"

steven-johnson commented:
Stil waiting for @alexreinking to approve since there are pending "requested changes"

alexreinking commented on src/CMakeLists.txt:
```diff
@@ -418,6 +423,91 @@ add_library(Halide
+        # fail on case-sensitive file systems. We'll try this as a hack workaround:
+        # just try all three. (Note that the internal CMake library name appears to be
+        # `flatbuffers` in all cases.)
+        set(FB_NAME "")
```
Have you considered using the NAMES argument to find_package? That would simplify this considerably

steven-johnson replied:
...because I didn't know about it?

alexreinking replied:
Sorry, sometimes I forget the abyss stared back into me

alexreinking commented on src/CMakeLists.txt:
```diff
@@ -418,6 +423,91 @@ add_library(Halide
+        VERBATIM
+    )
+    add_custom_target(generate_fb_header DEPENDS "${fb_header}")
+    set_source_files_properties("${fb_header}" PROPERTIES GENERATED TRUE)
```
Why is this necessary? Add custom command sets it automatically

steven-johnson replied:
I didn't realize that.

alexreinking commented on src/CMakeLists.txt:
```diff
@@ -418,6 +423,91 @@ add_library(Halide
+
+# flatbuffers is small and compiles quickly, but if you want/need to use
+# a local version (via find_package), configure with FLATBUFFERS_USE_FETCHCONTENT=OFF
+option(FLATBUFFERS_USE_FETCHCONTENT "Enable to download the Flatbuffers library via FetchContent" ON)
```
This whole FetchContent thing makes me sad. We'll (I'll) be able to do _much_ better with CMake 3.24+ and its tighter integration between the two methods.

alexreinking replied:
Actually, an upgrade would give us so much leverage that I think we should poll our users to see if they can tolerate an early upgrade to 3.24.

steven-johnson replied:
SGTM

alexreinking reviewed (approved):
On second thought, I can fix the NAMES thing in a follow up. Let's land this.

TH3CHARLie commented:
Thanks everyone for the help!

abadams commented on src/Pipeline.h:
```diff
@@ -184,11 +184,17 @@ class Pipeline {
      * outputs. Schedules the Funcs compute_root(). */
     Pipeline(const std::vector<Func> &outputs);
 
+    /** Make a pipeline from deserialization. */
```
This comment is confusing, because the args are not a "deserialization". A better comment would be something like: "Make a Pipeline with a vector of outputs and a list of requirement assertions. Used by deserialization."

abadams commented on src/Reduction.h:
```diff
@@ -54,6 +54,9 @@ class ReductionDomain {
      * the vector being outermost. */
     ReductionDomain(const std::vector<ReductionVariable> &domain);
 
+    /** Construct a reduction domain from deserialization */
```
This comment isn't very good either

abadams commented on src/Schedule.h:
```diff
@@ -652,6 +669,10 @@ class StageSchedule {
     }
     StageSchedule(const StageSchedule &other) = default;
     StageSchedule();
+    StageSchedule(const std::vector<ReductionVariable> &rvars, const std::vector<Split> &splits,
```
Needs a comment explaining what this constructor is for

abadams commented on src/Serialization.h:
```diff
@@ -0,0 +1,12 @@
+
+namespace Halide {
+
+void serialize_pipeline(const Pipeline &pipeline, const std::string &filename, std::map<std::string, Internal::Parameter> &params);
```
Badly needs a long comment explaining serialization.

abadams commented on src/Deserialization.h:
```diff
@@ -0,0 +1,22 @@
+namespace Halide {
+
+/**
+ * Deserialize a Halide pipeline from a file.
```
This comment needs to be longer, explaining in more detail what the map is for.

abadams commented on src/Deserialization.h:
```diff
@@ -0,0 +1,22 @@
+ */
+Pipeline deserialize_pipeline(const std::string &filename, const std::map<std::string, Internal::Parameter> &external_params);
+
+Pipeline deserialize_pipeline(std::istream &in, const std::map<std::string, Internal::Parameter> &external_params);
```
This function needs a doc string

abadams commented on src/Parameter.h:
```diff
@@ -55,6 +60,15 @@ class Parameter {
+              const Buffer<void> &buffer, int host_alignment, const std::vector<BufferConstraint> &buffer_constraints,
+              MemoryType memory_type);
+
+    Parameter(const Type &t, bool is_buffer, int dimensions, const std::string &name,
```
Needs a doc string. Note that in doxygen output this won't necessarily appear in a way that makes it clear that the comment on the previous constructor also applies.

TH3CHARLie commented:
I can send a PR today adding the JIT test we talked about in the dev meeting, enclosed in #ifdef blocks. Some docstring are addressed in Derek's recent PR #7760
